### PR TITLE
sd-journal: cleanups for generic_array_bisect_plus_one()

### PR DESCRIPTION
**Pull request by yuwata**
No functional changes, just refactoring.

split-out of #29334.

### CODE REVIEW

File: src/libsystemd/sd-journal/journal-file.c [hunk 1]
```diff
@@ -3107,46 +3107,37 @@ static int generic_array_bisect_plus_one(
...
-                step_back = direction == DIRECTION_UP;
-
-        if (r == TEST_RIGHT) {
-                if (direction == DIRECTION_DOWN)
-                        goto found;
-                else
+        /* If we are looking with DIRECTION_UP then we need to first see if in the actual array there is a
+         * matching entry, and return the last one of that. But if there isn't any we need to return this
+         * one. Hence remember this, and return it below. */
+        if (direction == DIRECTION_DOWN) {
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c:3122]: Can you add a comment for each of the scenarios we handle here as well? I'm having a hard time following

File: src/libsystemd/sd-journal/journal-file.c [hunk 2]
```diff
@@ -3550,7 +3550,7 @@ int journal_file_move_to_entry_by_monotonic_for_data(
 
         /* And now, continue seeking until we find an entry that exists in both bisection arrays. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c:3589]: Not exactly related to your changes, but I'm super confused by this comment, why do we need to find an entry that exists in both bisection arrays? Can you expand this comment a bit so I understand what's going on in this function? That would help me to review the changes.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Journal changes are always hard to review, can you add some more comments where I asked? That would help review the changes

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for the review. I added more comments in the code. Please take a look.

File: src/libsystemd/sd-journal/journal-file.c [hunk 3]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
-
-        /* if we are looking with DIRECTION_UP then we need to first
-           see if in the actual array there is a matching entry, and
-           return the last one of that. But if there isn't any we need
-           to return this one. Hence remember this, and return it
-           below. */
-        if (r == TEST_LEFT)
-                step_back = direction == DIRECTION_UP;
+        if (direction == DIRECTION_DOWN) {
+                /* If we are goind downwards, then we need to return the first object that passes the test.
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: goind => goind

File: src/libsystemd/sd-journal/journal-file.c [hunk 4]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
+                 * When there is no object that passes the test, we need to return the the first object that
+                 * test_object() returns TEST_RIGHT. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "we need to return the the first object that test_object() returns TEST_RIGHT" Seems there's something missing in this sentence?

File: src/libsystemd/sd-journal/journal-file.c [hunk 5]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
+                if (IN_SET(r,
+                           TEST_FOUND,  /* The 'extra' object passes the test. Hence, this is the first
+                                         * object that passes the test. */
+                           TEST_RIGHT)) /* The 'extra' object is the first object that test_object() returns
+                                           TEST_RIGHT, and no object exsits that passes the test. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "... returns TEST_RIGHT for ..."

Comment (daandemeyer) [reply src/libsystemd/sd-journal/journal-file.c]: "exists"

File: src/libsystemd/sd-journal/journal-file.c [hunk 6]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
+                        goto use_extra; /* The 'extra' object is exactly the one we are finding. It is not
+                                         * necessary to bisect the chained arrays. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "finding" => "looking for"

File: src/libsystemd/sd-journal/journal-file.c [hunk 7]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
+
+                /* Otherwise, the 'extra' object is not the one we are finding now. Search in the array. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "finding" => "looking for"

File: src/libsystemd/sd-journal/journal-file.c [hunk 8]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
 
-        if (r == TEST_RIGHT) {
-                if (direction == DIRECTION_DOWN)
-                        goto found;
-                else
-                        return 0;
+        } else {
+                /* If we are goind upwards, then we need to return the last object that passes the test.
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "going"

File: src/libsystemd/sd-journal/journal-file.c [hunk 9]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
+                 * When there is no object that passes the test, we need to return the the last object that
+                 * test_object() returns TEST_LEFT. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "returns TEST_LEFT for"

File: src/libsystemd/sd-journal/journal-file.c [hunk 10]
```diff
@@ -3107,46 +3107,56 @@ static int generic_array_bisect_plus_one(
...
+                if (r == TEST_RIGHT)
+                        return 0; /* The 'extra' object and all objects in the chained arrays will never
+                                   * return TEST_FOUND or TEST_LEFT. There is no object we are finding. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/journal-file.c]: "The object we are looking for does not exist"

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for the review. The suggestions are applied. Upgrading the green label.

**Pull request merged**